fix(): Remove redundant indexes to resolve schema warnings

DDA1O1 opened the pull request:
Fixes #13735

### What

This Pull Request introduces new database migrations to remove multiple redundant indexes across several core modules, including product, cart, order, customer, and inventory.

### Why

As detailed in issue #13735, a fresh Medusa installation produces numerous "Duplicate Index" warnings. These legacy indexes add unnecessary write overhead and provide no performance benefit. This PR cleans up the schema to resolve these warnings and improve database health.

### How

I have added one new, reversible migration file for each of the five affected modules:
- `@medusajs/product`
- `@medusajs/cart`
- `@medusajs/order`
- `@medusajs/customer`
- `@medusajs/inventory`

Each migration's `up()` method safely drops the older, redundant index, and its `down()` method re-creates it, ensuring the change is fully reversible and non-destructive.


### Testing

I have tested these migrations by following the local development workflow outlined in the `CONTRIBUTING.md` guide.

1.  **Setup:**
    *   Cloned my forked Medusa repository locally .
    *   Created a separate, fresh test project using `npx create-medusa-app@latest my-medusa-store`.
    *   The test project's PostgreSQL database, which already contained the schema with the duplicate indexes.

2.  **Linking Local Source Code:**
    *   In the test project's `package.json`, I modified all `@medusajs/*` dependencies and resolutions to point to the local packages in my forked repository (e.g., `"@medusajs/product": "file:../medusa/packages/modules/product"`).
    *   From the test project's directory, I ran `yarn install` to link the local, modified Medusa source code into its `node_modules`.

3.  **Build & Migration:**
    *   Inside my forked Medusa repository, I ran `yarn build` to compile the new TypeScript migration files.
    *   From the root of the **test project**, I then executed the migration command: `npx medusa migration run`.

4.  **Verification:**
    *   The command successfully identified and ran only the new migration files I had created.
    *   I also confirmed via direct SQL queries that the old, redundant indexes were correctly dropped from all affected tables (`product_collection`, `customer_group`, etc.).

DDA1O1 committed:
fix(db): Resolve duplicate index on product_collection

DDA1O1 committed:
fix(db): Remove redundant indexes on cart-related tables

Several tables within the cart module had duplicate indexes created by older migrations. These provide no performance benefit and add unnecessary overhead to write operations (INSERT, UPDATE).

This commit introduces a new migration to safely drop the older, redundant indexes, keeping the more explicitly named ones.

Removes the following indexes:
- `IDX_line_item_cart_id` from `cart_line_item`
- `IDX_adjustment_item_id` from `cart_line_item_adjustment`
- `IDX_tax_line_item_id` from `cart_line_item_tax_line`
- `IDX_shipping_method_cart_id` from `cart_shipping_method`
- `IDX_adjustment_shipping_method_id` from `cart_shipping_method_adjustment`
- `IDX_tax_line_shipping_method_id` from `cart_shipping_method_tax_line`

DDA1O1 committed:
fix(db): Remove redundant indexes on order-related tables

Database linters identified several instances of duplicate indexes on the `order_item` and `order_shipping` tables.

DDA1O1 committed:
fix(db): Remove redundant unique index on customer_group table

An analysis of the database schema identified two functionally identical unique indexes on the `name` column of the `customer_group` table: `IDX_customer_group_name` and `IDX_customer_group_name_unique`.

DDA1O1 committed:
fix(db): Remove redundant composite index on inventory_level table

A database schema analysis revealed two functionally identical unique composite indexes on the `inventory_level` table: `IDX_inventory_level_item_location` and `IDX_inventory_level_location_id_inventory_item_id`.

DDA1O1 committed:
Merge branch 'medusajs:develop' into develop

willbouch commented on .changeset/poor-llamas-march.md:
```diff
@@ -0,0 +1,9 @@
+"@medusajs/product": patch
+---
+
+fix(db): Remove redundant indexes from product, cart, order, customer, and inventory modules to improve database schema health.
```
We usually put the name of the modules affected (db is not a module). In your case, the change affects too many of them to put there, so you can just write `fix()` and that would be fine :)

willbouch commented on packages/modules/order/src/migrations/Migration20251010125912.ts:
```diff
@@ -0,0 +1,39 @@
+  override async up(): Promise<void> {
+     // 1. On `order_item`: Drops the old composite index.
+    this.addSql(
+      'DROP INDEX IF EXISTS "IDX_order_item_order_id_version";'
```
Maybe we should delete `IDX_order_item_version` ? The one you are deleting right now seems to be the one with the most "correct" name

DDA1O1 replied:
hey i removed it cause the latest migration file introduced it so but we can go with the old one

willbouch commented on packages/modules/order/src/migrations/Migration20251010125912.ts:
```diff
@@ -0,0 +1,39 @@
+
+    // 2. On `order_shipping`: Drops the old composite index.
+    this.addSql(
+      'DROP INDEX IF EXISTS "IDX_order_shipping_order_id_version";'
```
Same here, I think we should drop the other one no?

DDA1O1 replied:
done

willbouch commented on packages/modules/product/src/migrations/Migration20251011090616.ts:
```diff
@@ -0,0 +1,18 @@
+// UP: Adds the feature (the correct index on product_category).
+  override async up(): Promise<void> {
+    this.addSql(
+      'CREATE INDEX IF NOT EXISTS "IDX_product_category_deleted_at" ON "product_category" ("deleted_at");'
```
Can you remove this? I am fine with removing indexes that are duplicates, but can we avoid adding new ones? I prefer keeping this a status quo for performance reasons. I am sure this would be ok, but I prefer to be safer. We have people working on performance right now in the team, so if they believe an index is missing here, they will add it :)

DDA1O1 replied:
sure

DDA1O1 committed:
Update poor-llamas-march.md

DDA1O1 committed:
Update Migration20251010125912.ts

DDA1O1 committed:
Delete packages/modules/product/src/migrations/Migration20251011090616.ts

DDA1O1 committed:
Merge branch 'develop' into fix/db-duplicate-indexes

DDA1O1 commented:
hey @willbouch , I've made all the changes you suggested.

willbouch committed:
Merge branch 'develop' into fix/db-duplicate-indexes

willbouch committed:
Merge branch 'develop' into fix/db-duplicate-indexes

willbouch committed:
Merge branch 'develop' into fix/db-duplicate-indexes

willbouch commented:
Hey @DDA1O1. It looks like those indexes (at least some I checked) are also defined in the models (so the TS files defining models like "Order" or "OrderShipping". For example, [here](https://github.com/medusajs/medusa/blob/823a5c75ff97cf0acbd4ddf00f43ab3b49f0f210/packages/modules/order/src/models/order-shipping-method.ts#L63-L68) 

WE should clean that up/make sure the stuff there is up to date

DDA1O1 commented:
> Hey @DDA1O1. It looks like those indexes (at least some I checked) are also defined in the models (so the TS files defining models like "Order" or "OrderShipping". For example, [here](https://github.com/medusajs/medusa/blob/823a5c75ff97cf0acbd4ddf00f43ab3b49f0f210/packages/modules/order/src/models/order-shipping-method.ts#L63-L68)
> 
> WE should clean that up/make sure the stuff there is up to date

hey thanks for letting me know definitely going to clean  the dropped indexes

DDA1O1 committed:
Merge branch 'medusajs:develop' into fix/db-duplicate-indexes

DDA1O1 committed:
Remove unused index from order shipping method model

Removed unused index 'IDX_order_shipping_version' from order shipping methods model.

DDA1O1 committed:
fix(): removed duplicate indexes from models as well

willbouch commented:
I see you cleaned them up, but the thing is that I feel some are missing no? You cleaned up duplicates, that's fine, but it looks like the model now do not define the IDX that actually live in the database. The model has one IDX defined, which you removed, but the database had 2 IDX and you removed one. That means that the model is missing the actual IDX missing no?

DDA1O1 commented:
I should have talked to you before implementing any change. Is there anything else that i should watch out or should i define  the indexes that actually live  in the database

willbouch commented:
@DDA1O1 Yeah I did not know before yesterday that the IDS were defined in the model too, sorry on that one. That being said, I believe you should just clean up the ones that are not in the model. Even if the name of the index is not the best one of the 2, it is simpler to get rid of those ones that are not in the model I think

DDA1O1 commented:
ok here is what i found for some indexes like in, order module the indexes are defined in models so i can replace them with the best name as you suggested. For modules like inventory, customer and product the indexes are defined in old migration files itself not in the models so it does not need any change.

DDA1O1 committed:
minor change

DDA1O1 commented on packages/modules/order/src/models/order-item.ts:
```diff
@@ -34,7 +34,7 @@ const _OrderItem = model
     },
     {
-      name: "IDX_order_item_version",
+      name: "IDX_order_item_order_id_version",
```
its fine

DDA1O1 committed:
fix: update indexes in order models to include order_id

DDA1O1 committed:
fix: update order_item and order_shipping indexes to include order_id

DDA1O1 committed:
bug bot

DDA1O1 commented:
hey @willbouch  can you check it out

willbouch commented on packages/modules/cart/src/models/line-item-adjustment.ts:
```diff
@@ -24,10 +24,10 @@ const LineItemAdjustment = model
+      name: "IDX_cart_line_item_adjustment_item_id",
       on: ["item_id"],
-      where: "deleted_at IS NULL",
+      where: "deleted_at IS NULL AND item_id IS NOT NULL",
```
I see you add something in the where here (and you do it for other models) but the migration file does not update the actual index in the DB? So, the model will say `where: "deleted_at IS NULL AND item_id IS NOT NULL"` but what we actually have in the DB is the already existing index we have which seems to only have `where: "deleted_at IS NULL"`. Or am I missing something?

DDA1O1 replied:
no you are right the definition should exactly match with the index definition that lives on the DB

DDA1O1 committed:
Merge branch 'develop' into fix/db-duplicate-indexes

DDA1O1 committed:
testing out if this works with orginal snapshot
